cpp/test/sg/fil_child_index_test.cu

```diff
@@ -38,28 +38,39 @@ struct proto_inner_node {
   bool is_categorical = false;  // see base_node::is_categorical
   int fid             = 0;      // feature id, see base_node::fid
   int set             = 0;      // which bit set represents the matching category list
-  float thresh        = 0.0f;   // threshold, see base_node::thresh
+  double thresh       = 0.0;    // threshold, see base_node::thresh
   int left            = 1;      // left child idx, see sparse_node*::left_index()
-  val_t<float> split()
+  template <typename real_t>
+  val_t<real_t> split()
   {
-    val_t<float> split;
+    val_t<real_t> split;
     if (is_categorical)
       split.idx = set;
+    else if (std::isnan(thresh))
+      split.f = std::numeric_limits<real_t>::quiet_NaN();
     else
-      split.f = thresh;
+      split.f = static_cast<real_t>(thresh);
     return split;
   }
+  operator dense_node<float>()
+  {
+    return dense_node<float>({}, split<float>(), fid, def_left, false, is_categorical);
+  }
+  operator dense_node<double>()
+  {
+    return dense_node<double>({}, split<double>(), fid, def_left, false, is_categorical);
+  }
   operator sparse_node16<float>()
   {
-    return sparse_node16<float>({}, split(), fid, def_left, false, is_categorical, left);
+    return sparse_node16<float>({}, split<float>(), fid, def_left, false, is_categorical, left);
   }
-  operator sparse_node8()
+  operator sparse_node16<double>()
   {
-    return sparse_node8({}, split(), fid, def_left, false, is_categorical, left);
+    return sparse_node16<double>({}, split<double>(), fid, def_left, false, is_categorical, left);
   }
```

Comment:
Same here: Any reason not to use a single template for this?

Reply:
Done.